disperser/meterer/meterer.go

```diff
@@ -0,0 +1,273 @@
+		return fmt.Errorf("failed to update cumulative payment: %w", err)
+	}
+	// Validate payments attached
+	err = m.ValidatePayment(ctx, blobHeader, onDemandPayment)
+	if err != nil {
+		fmt.Println("validation failed: %w", err)
+		//TODO: Make sure this remove always works or handle the error
+		m.OffchainStore.RemoveOnDemandPayment(ctx, blobHeader.AccountID, blobHeader.CumulativePayment)
+		return fmt.Errorf("invalid on-demand payment: %w", err)
+	}
+
+	// Update bin usage atomically and check against bin capacity
+	if err := m.IncrementGlobalBinUsage(ctx, blobHeader, m.GlobalBytesPerSecond); err != nil {
+		//TODO: conditionally remove the payment based on the error type (maybe if the error is store-op related)
+		m.OffchainStore.RemoveOnDemandPayment(ctx, blobHeader.AccountID, blobHeader.CumulativePayment)
+		return fmt.Errorf("failed global rate limiting")
+	}
+
+	return nil
+}
+
+// ValidatePayment checks if the provided payment header is valid against the local accounting
+// prevPmt is the largest  cumulative payment strictly less    than blobHeader.cumulativePayment if exists
+// nextPmt is the smallest cumulative payment strictly greater than blobHeader.cumulativePayment if exists
+// nextPmtBlobSize is the blobSize of corresponding to nextPmt if exists
+func (m *Meterer) ValidatePayment(ctx context.Context, blobHeader BlobHeader, onDemandPayment *OnDemandPayment) error {
+
+	prevPmt, nextPmt, nextPmtBlobSize, err := m.OffchainStore.GetRelevantOnDemandRecords(ctx, blobHeader.AccountID, blobHeader.CumulativePayment) // zero if DNE
+	if err != nil {
+		return fmt.Errorf("failed to get relevant on-demand records: %w", err)
+	}
+	// the current request must increment cumulative payment by a magnitude sufficient to cover the blob size
+	if prevPmt+uint64(blobHeader.BlobSize)*m.Config.PricePerByte > blobHeader.CumulativePayment {
+		return fmt.Errorf("insufficient cumulative payment increment")
+	}
+	// the current request must not break the payment magnitude for the next payment if the two requests were delivered out-of-order
+	if nextPmt != 0 && blobHeader.CumulativePayment+uint64(nextPmtBlobSize)*m.Config.PricePerByte > nextPmt {
+		return fmt.Errorf("breaking cumulative payment invariants")
+	}
+	// check passed: blob can be safely inserted into the set of payments
+	// prevPmt + blobHeader.BlobSize * m.FixedFeePerByte <= blobHeader.CumulativePayment
+	//                                                   <= nextPmt - nextPmtBlobSize * m.FixedFeePerByte > nextPmt
+	return nil
+}
+
+// ValidateBinIndex checks if the provided bin index is valid
+func (m *Meterer) ValidateGlobalBinIndex(blobHeader BlobHeader) (uint64, error) {
+	// Deterministic function: local clock -> index (1second intervals)
+	currentBinIndex := uint64(time.Now().Unix())
+
+	// Valid bin indexes are either the current bin or the previous bin (allow this second or prev sec)
+	if blobHeader.BinIndex != currentBinIndex && blobHeader.BinIndex != (currentBinIndex-1) {
```

Comment:
For global rate limit, I think we can argue that we don't need to include the BinIndex within the header, but just use the receipt time directly.

Reply:
makes sense, in this case we don't need to validate the bin index, maybe not even receipt time, but the time when global rate gets checked (since global rate limit check happens after payment validations, we avoid failing on cases where payment validation ever gets to more than 2 seconds)